cmd/thanos/main.go

```diff
@@ -211,3 +224,21 @@ func interrupt(logger log.Logger, cancel <-chan struct{}) error {
 		return errors.New("canceled")
 	}
 }
+
+func reload(logger log.Logger, cancel <-chan struct{}, r chan<- struct{}) error {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, syscall.SIGHUP)
+	for {
+		select {
+		case s := <-c:
+			level.Info(logger).Log("msg", "caught signal. Reloading.", "signal", s)
+			select {
+			case r <- struct{}{}:
```

Comment:
What if no one is receiving here?

Reply:
I think we mind this 
```
{
		cancel := make(chan struct{})
		g.Add(func() error {
			return reload(logger, cancel, reloadCh)
		}, func(error) {
			close(cancel)
		})
	}
```
To be per command maybe? Or noop otherwise? That's the reason and also we might think of redesigning commands and the number of arguments is becoming a problem (:

Reply:
> What if no one is receiving here?

When no one is receiving that means (or if someone's receiving), we just allocate 4 bytes of memory for the full lifecycle of the program. It's unnecessary if no one receiving of course. Empty structs allocate zero bytes of memory, fortunately. (so allocation is just for the channel)
https://play.golang.org/p/2QPjVBQZSGU

That being said I'll have another look at it to avoid unnecessary allocation.

> To be per command maybe? Or noop otherwise? 

I will have another look at this, find a clean way to achieve it.
I just don't want to handle signals outside the `main.go`, it makes things harder to pinpoint.

> That's the reason and also we might think of redesigning commands and the number of arguments is becoming a problem (:

Yes, we definitely need to find more elegant solution to this parameter problem. As once a great man said:

> “If you have a procedure with 10 parameters, you probably missed some.”
> —Allan Perlis

🤪

Reply:
> When no one is receiving that means (or if someone's receiving), we just allocate 4 bytes of memory for the full lifecycle of the program

I don't think only 4 bytes. I am more worried on signal channel.Whatt if no one is receiving and I will send 100mln of SIGHUP signals?

Reply:
For redesign we probably need to finally move to some structure (: I remember Cortex had some quite nice modules structure.. worth to get inspired? Cc @gouthamve

Reply:
> Whatt if no one is receiving and I will send 100mln of SIGHUP signals?
That’s why it’s in an additional select with default case so that it can just continue on if the buffered (reload) channel is full. So memory will always be constant, it’ll just loop over 100mln.

Reply:
You are right, I missed this default case. Thanks! ((: